File: iq/20/number/is_happy_number.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package number

import "iter"

const happyNumEnd = 1

// IsHappyNumber reports if the number n is "happy", i.e., iterative sum of
// squares of its digits leads to 1.
func IsHappyNumber(n int) bool {
	for slow, fast := range runNumbers(n) {
		if fast == happyNumEnd {
			break
		}
		if slow == fast {
			return false
		}
	}
	return true
}

func runNumbers(n int) iter.Seq2[int, int] {
	return func(yield func(int, int) bool) {
		if n <= happyNumEnd {
			return
		}
		slow, fast := n, getNextNumber(n)
		for fast != happyNumEnd {
			if !yield(slow, fast) {
				break
			}
			slow = getNextNumber(slow)
			fast = getNextNumber(fast)
			fast = getNextNumber(fast)
		}
	}
}

func getNextNumber(n int) int {
	next := 0
	for ; n > 0; n /= 10 {
		i := n % 10
		if i == 0 {
			continue
		}
		next += i * i
	}
	return next
}

File: iq/20/number/is_happy_number.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// clang-format off-next-line
//go:build ignore

#include "iq/20/number/is_happy_number.h"

namespace iq::number {
namespace {

int getNextNumber(int n) {
  int next = 0;
  for (; n != 0; n /= 10) {
    const int k = n % 10;
    if (k == 0) {
      continue;
    }
    next += k * k;
  }
  return next;
}

}  // namespace

bool IsHappyNumber(int n) {
  int slow = n;
  int fast = getNextNumber(n);
  while (fast != 1) {
    if (slow == fast) {
      return false;
    }
    slow = getNextNumber(slow);
    fast = getNextNumber(fast);
    fast = getNextNumber(fast);
  }
  return true;
}

}  // namespace iq::number
